tools/rpkg/src/statement.cpp

```diff
@@ -268,20 +274,67 @@ static void transform(Vector &src_vec, SEXP &dest, idx_t dest_offset, idx_t n) {
 	case LogicalTypeId::INTEGER:
 		VectorToR<int32_t, uint32_t>(src_vec, n, INTEGER_POINTER(dest), dest_offset, NA_INTEGER);
 		break;
+	case LogicalTypeId::TIMESTAMP_SEC: {
+		auto src_data = FlatVector::GetData<int64_t>(src_vec);
+		auto &mask = FlatVector::Validity(src_vec);
+		double *dest_ptr = ((double *)NUMERIC_POINTER(dest)) + dest_offset;
+		for (size_t row_idx = 0; row_idx < n; row_idx++) {
+			dest_ptr[row_idx] = !mask.RowIsValid(row_idx) ? NA_REAL : (double)src_data[row_idx];
+		}
+
+		// some dresssup for R
+		SET_CLASS(dest, RStrings::get().POSIXct_POSIXt_str);
+		Rf_setAttrib(dest, RStrings::get().tzone_sym, RStrings::get().UTC_str);
+		break;
+	}
+	case LogicalTypeId::TIMESTAMP_MS: {
+		auto src_data = FlatVector::GetData<int64_t>(src_vec);
+		auto &mask = FlatVector::Validity(src_vec);
+		double *dest_ptr = ((double *)NUMERIC_POINTER(dest)) + dest_offset;
+		for (size_t row_idx = 0; row_idx < n; row_idx++) {
+			dest_ptr[row_idx] =
+			    !mask.RowIsValid(row_idx) ? NA_REAL : (double)src_data[row_idx] / Interval::MSECS_PER_SEC;
+		}
+
+		// some dresssup for R
+		SET_CLASS(dest, RStrings::get().POSIXct_POSIXt_str);
+		Rf_setAttrib(dest, RStrings::get().tzone_sym, RStrings::get().UTC_str);
+		break;
+	}
 	case LogicalTypeId::TIMESTAMP: {
-		auto src_data = FlatVector::GetData<timestamp_t>(src_vec);
+		auto src_data = FlatVector::GetData<int64_t>(src_vec);
 		auto &mask = FlatVector::Validity(src_vec);
 		double *dest_ptr = ((double *)NUMERIC_POINTER(dest)) + dest_offset;
 		for (size_t row_idx = 0; row_idx < n; row_idx++) {
 			dest_ptr[row_idx] =
-			    !mask.RowIsValid(row_idx) ? NA_REAL : (double)Timestamp::GetEpochSeconds(src_data[row_idx]);
+			    !mask.RowIsValid(row_idx) ? NA_REAL : (double)src_data[row_idx] / Interval::MICROS_PER_SEC;
 		}
 
 		// some dresssup for R
 		SET_CLASS(dest, RStrings::get().POSIXct_POSIXt_str);
 		Rf_setAttrib(dest, RStrings::get().tzone_sym, RStrings::get().UTC_str);
 		break;
 	}
+	case LogicalTypeId::TIMESTAMP_NS: {
+		auto src_data = FlatVector::GetData<int64_t>(src_vec);
+		auto &mask = FlatVector::Validity(src_vec);
+		double *dest_ptr = ((double *)NUMERIC_POINTER(dest)) + dest_offset;
+		for (size_t row_idx = 0; row_idx < n; row_idx++) {
+			dest_ptr[row_idx] =
+			    !mask.RowIsValid(row_idx) ? NA_REAL : (double)src_data[row_idx] / Interval::NANOS_PER_SEC;
+		}
+
+		// some dresssup for R
+		SET_CLASS(dest, RStrings::get().POSIXct_POSIXt_str);
+		Rf_setAttrib(dest, RStrings::get().tzone_sym, RStrings::get().UTC_str);
+
+		static bool precision_warning = false;
```

Comment:
This isn't thread safe, right? Does it matter or is this code guaranteed to only be called once because of the R interpreter? @hannesmuehleisen thoughts on this?

Reply:
Could use [`std::call_once`](https://en.cppreference.com/w/cpp/thread/call_once) for this.